File: ruby/word-count/word_count.rb
class Phrase
  attr_accessor :words, :counts

  def initialize(sentence)
    @words = sentence
    @counts = Hash.new(0)
  end

  def word_count
    sanitize
    words.each { |word| counts[word] += 1 }
    counts
  end

  def sanitize
    @words = words
             .downcase
             .gsub(/[^\w']/, ' ')    # replace with  space, except quotes
             .gsub(/'(\w+)'/, '\1')  # remove outer quotes
             .squeeze(' ')
             .split(' ')
  end
end

File: elixir/word-count/word_count.exs
defmodule Words do
  @spec count(String.t) :: map
  def count(sentence) do
    String.downcase(sentence)
    |> String.replace(~r/[^\w-]|_/u, " ")
    |> String.split
    |> Enum.group_by(&(&1))
    |> Enum.map(fn({word, counts}) -> { word, length(counts) } end)
    |> Enum.into(%{})
  end
end
